File: python/utils/leitor_csv.py
import csv
from datetime import datetime
from typing import Callable
from model.gerenciador_dados import GerenciadorDados
from model.partida import Partida

class LeitorCSV:
    @staticmethod
    def carregar_dados_completos(caminho_arquivo: str, gerenciador: GerenciadorDados, 
                               callback: Callable[[int, str], None]):
        """Carrega dados completos de um arquivo CSV"""
        try:
            callback(0, "Lendo arquivo...")
            
            with open(caminho_arquivo, 'r', encoding='utf-8') as arquivo:
                # Contar linhas para progresso
                total_linhas = sum(1 for _ in arquivo) - 1  # -1 para descontar cabeçalho
                
            callback(10, "Processando dados...")
            
            linhas_processadas = 0
            
            with open(caminho_arquivo, 'r', encoding='utf-8') as arquivo:
                reader = csv.reader(arquivo)
                headers = next(reader)  # Pular cabeçalho
                
                # Validar cabeçalho
                if len(headers) < 7:
                    raise Exception("Formato de arquivo inválido. Esperado: ano,rodada,data,mandante,visitante,gols_mandante,gols_visitante")
                
                for linha in reader:
                    try:
                        # Parsing dos dados
                        ano = int(linha[0].strip())
                        rodada = int(linha[1].strip())
                        data = datetime.strptime(linha[2].strip(), '%d/%m/%Y').date()
                        mandante = linha[3].strip()
                        visitante = linha[4].strip()
                        gols_mandante = int(linha[5].strip())
                        gols_visitante = int(linha[6].strip())
                        
                        partida = Partida(ano, rodada, data, mandante, visitante,
                                        gols_mandante, gols_visitante)
                        gerenciador.adicionar_partida(partida)
                        
                        linhas_processadas += 1
                        
                        # Atualizar progresso a cada 100 linhas
                        if linhas_processadas % 100 == 0:
                            progresso = min(90, 10 + (linhas_processadas * 80 // total_linhas))
                            callback(progresso, f"Processadas {linhas_processadas} partidas...")
                    
                    except Exception as ex:
                        print(f"Erro na linha {linhas_processadas + 2}: {ex}")
                        continue
            
            callback(95, "Finalizando...")
            
            if linhas_processadas == 0:
                raise Exception("Nenhuma partida válida foi encontrada no arquivo.")
            
            callback(100, f"Carregamento concluído! {linhas_processadas} partidas processadas.")
            
        except FileNotFoundError:
            raise Exception(f"Arquivo não encontrado: {caminho_arquivo}")
        except Exception as e:
            raise Exception(f"Erro ao processar arquivo: {str(e)}")

File: java/src/main/java/utils/LeitorCSV.java
package utils;

import model.GerenciadorDados;
import model.Partida;
import com.opencsv.CSVReader;

import java.io.FileReader;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.function.BiConsumer;

public class LeitorCSV {
    private static final DateTimeFormatter DATE_FORMATTER = DateTimeFormatter.ofPattern("dd/MM/yyyy");
    
    @FunctionalInterface
    public interface ProgressCallback {
        void onProgress(int progresso, String mensagem);
    }
    
    public static void carregarDadosCompletos(String caminhoArquivo, GerenciadorDados gerenciador, 
                                            ProgressCallback callback) throws Exception {
        
        callback.onProgress(0, "Lendo arquivo...");
        
        try (CSVReader reader = new CSVReader(new FileReader(caminhoArquivo))) {
            String[] headers = reader.readNext(); // Ler cabeçalho
            
            // Validar cabeçalho esperado
            if (headers == null || headers.length < 7) {
                throw new Exception("Formato de arquivo inválido. Esperado: ano,rodada,data,mandante,visitante,gols_mandante,gols_visitante");
            }
            
            String[] linha;
            int totalLinhas = 0;
            int linhasProcessadas = 0;
            
            // Primeira passada para contar linhas
            callback.onProgress(10, "Contando registros...");
            while ((linha = reader.readNext()) != null) {
                totalLinhas++;
            }
            
            // Segunda passada para processar dados
            callback.onProgress(20, "Processando dados...");
            
        } catch (Exception e) {
            // Reabrir arquivo para processamento real
        }
        
        // Processar arquivo novamente
        try (CSVReader reader = new CSVReader(new FileReader(caminhoArquivo))) {
            String[] headers = reader.readNext(); // Pular cabeçalho
            String[] linha;
            int linhasProcessadas = 0;
            int totalEstimado = 10000; // Estimativa para progresso
            
            while ((linha = reader.readNext()) != null) {
                try {
                    // Parsing dos dados - formato esperado:
                    // ano,rodada,data,mandante,visitante,gols_mandante,gols_visitante
                    int ano = Integer.parseInt(linha[0].trim());
                    int rodada = Integer.parseInt(linha[1].trim());
                    LocalDate data = LocalDate.parse(linha[2].trim(), DATE_FORMATTER);
                    String mandante = linha[3].trim();
                    String visitante = linha[4].trim();
                    int golsMandante = Integer.parseInt(linha[5].trim());
                    int golsVisitante = Integer.parseInt(linha[6].trim());
                    
                    Partida partida = new Partida(ano, rodada, data, mandante, visitante, 
                                                golsMandante, golsVisitante);
                    gerenciador.adicionarPartida(partida);
                    
                    linhasProcessadas++;
                    
                    // Atualizar progresso a cada 100 linhas
                    if (linhasProcessadas % 100 == 0) {
                        int progresso = Math.min(90, 20 + (linhasProcessadas * 70 / totalEstimado));
                        callback.onProgress(progresso, 
                            String.format("Processadas %d partidas...", linhasProcessadas));
                    }
                    
                } catch (Exception ex) {
                    // Log do erro, mas continua processamento
                    System.err.println("Erro na linha " + (linhasProcessadas + 2) + ": " + ex.getMessage());
                }
            }
            
            callback.onProgress(95, "Finalizando...");
            
            if (linhasProcessadas == 0) {
                throw new Exception("Nenhuma partida válida foi encontrada no arquivo.");
            }
            
            callback.onProgress(100, String.format("Carregamento concluído! %d partidas processadas.", 
                                                  linhasProcessadas));
        }
    }
}